### DISCUSSION TITLE
Add possibility to define nonlinear constraints behaviour

### DISCUSSION THREAD

**Discussion by sonkaste17** [StructuralAnalysisFormat/StructuralAnalysisFormat-Doc#95] [Ideas]
Store nonlinear curves for scaffolding couplers, supports etc.
To be defined.

**Comment by vojtech-chalupa** [comment]
Hi, +1 on this from us at IDEA. The workflow (3rd party software -> SAF -> IDEA) of some of our customers (those using nonlinear constraints in their models) is currently blocked because SAF can't store this data.

**Comment by MartinSalak** [comment]
updated 24.10.2022 - Stiffness requirement changed

Here is a proposal:
We would like to create a new object called NonlinearFunction to store the nonlinear properties. Please, have a look SAF community. Below you can see proposed changes for the existing object RelConnectsStructuralMember and new object NonlinearFunction. All remarks are welcome.

Existing object: **RelConnectsStructuralMember**
Changed rows:
| Property | Example | Definition |
| --- | --- | --- |
| ux | Free, Rigid, Flexible, **Nonlinear** | Translation in X direction. Free - That is it imposes no constraint in the direction. Rigid - The connection in fully rigid in the specified direction. Flexible - The connection is flexible (elastic) in the specified direction. ~~Parameter Flexible can be linear only, non-linearity is not supported.~~ **Nonlinear – The connection’s behavior is set by a function** |
| uy |  |  |
| uz |  |  |
| fix |  |  |
| fiy |  |  |
| fiz |  |  |
| Stiffness X [MN/m] | 100 | Now required for ux = Flexible **or Nonlinear**. For Nonlinear it will be used for linear analysis and it can be used as a starting point for nonlinear analysis.|
| Stiffness Y [MN/m] | | |
| Stiffness Z [MN/m] | | |
| Stiffness Fix [MNm/rad] | | |
| Stiffness Fiy [MNm/rad] | | |
| Stiffness Fiz [MNm/rad] | | |

New rows:
| Property | Example | Definition |
| --- | --- | --- |
| Function ux | Scaffolding coupler 1 | Reference to object NonlinearFunction, applicable if ux = Nonlinear. Nonlinear function must be of type Translation |
| Function uy | | |
| Function uz | | |
| Function fix | | |
| Function fiy | | |
| Function fiz | | |


New object: **NonlinearFunction**
Nonlinear function defines the change of stiffness based on the deformation. It is a definition of a graph where the horizontal x-axis is the deformation and the vertical y-axis is the force or moment caused by the deformation (translation or rotation).
![image](https://user-images.githubusercontent.com/89633938/195854092-d7656304-ea0d-4ce6-a3df-24e998ebf873.png)
 
| Property | Example | Definition |
| --- | --- | --- |
| Name | Scaffolding coupler 1 | Unique name |
| Type | Translation, Rotation | Enum which sets the units for the function. Translation is in [m, MN]. Rotation is in [rad, kNm].  |
| Translation function |-0.05; -1\| 0; 0\| 0.05; 5.0\| 0.1; 10.0| Coordinates of a function for translational stiffness in \|[m]; [MN]\|.<br>Recommended maximum value for y axis is 1e10, anything above should be considered rigid. The format of the data is: x1; y1\|x2; y2\|x3; y3\|xi; yi\|xi+1; yi+1.The coordinates of the points are divided by ";" and each point is separated by "\|". Points should be written in order from left to right on the x axis. For each translation or rotation there can be only one force or moment – it is not possible to draw vertical lines. The minimum distance between two consecutive points must not be smaller than 1e-6.<br>X and Y coordinates of one point must be either both positive or negative.<br>The graph must go through the origin 0; 0. |
| Rotation function |-0.05; -1\| 0; 0\| 0.05; 5.0\| 0.1; 10.0| Coordinates of a function for rotational stiffness \|[rad]; [kNm]\|.<br>Recommended maximum value for y axis is 1e10, anything above should be considered rigid. The format of the data is: x1; y1\|x2; y2\|x3; y3\|xi; yi\|xi+1; yi+1.The coordinates of the points are divided by ";" and each point is separated by "\|". Points should be written in order from left to right on the x axis. For each translation or rotation there can be only one force or moment – it is not possible to draw vertical lines. The minimum distance between two consecutive points must not be smaller than 1e-6.<br>X and Y coordinates of one point must be either both positive or negative.<br>The graph must go through the origin 0; 0. |
| Positive end | Rigid, Free, Flexible | This enum defines the continuation of the function after the last point in the graph. 'Rigid' means vertical end of function, 'Free' means horizontal end of function and 'Flexible' means continuing of function in same direction of last part of function. | 
| Negative end | Rigid, Free, Flexible |  This enum defines the continuation of the function before the first point in the graph. 'Rigid' means vertical end of function, 'Free' means horizontal end of function and 'Flexible' means continuing of function in same direction of last part of function. |

Notes:
Nonlinear function can be displayed in a graph, look at the example below to see the difference between Rigid, Free and Flexible ends.  
 
![image](https://user-images.githubusercontent.com/89633938/195854105-842f04bc-e1da-4f0e-b642-c094b86f7bbd.png)

It is not possible to draw vertical lines. Use rigid positive or negative ends to create rigid behavior or crate a line that is almost vertical.

**Comment by dschuermans** [comment]
For RelConnectsStructuralMember i don't see any issues, provided I understand the changes correctly:
- Allow `Nonlinear` as value for type which is currently being validated against
- Update validation for stiffness values to be required also when type is `Nonlinear`
- Add 6 new **optional?** reference properties, which reference the new object
  Question though: is the reference required when corresponding constraint type is set to `Nonlinear`?

Concerning the new object, I have the following remark:
- You expect to have a single column in which you can define multiple values separated by multiple separators. **But they can also be a different unit, depending on the type of the function**
  
  Maybe split this up in "Translation function" and "Rotation function" columns? 
  Because the SDK doesn't support the scenario you're trying to achieve

  Unless I'm missing something here, do we already have an object in which we have such behavior?
- Should we validate the values and not allow values greater then 1e10?
- I'm not sure what is meant with `For each translation or rotation there can be only one force or moment`

**Reply by MartinSalak** [reply]
Thanks for the remarks @dschuermans :)

For RelConnectsStructuralMember:
- yes, `Nonlinear` is the new value for enums ux, uy, uz, fix, fiy, fiz
- yes `Stiffness` is required for Nonlinear as well (so it is possible to still compute it linearly)
- yes for `Nonlinear` constraint type it is required to provide NonlinearFunction reference

For the new NonlinearFunction:
- I have split it, do you like it like this?
- no, values greater than 1e10 should not be validated, I modified the text (we should though change values greater than 1e10 to 1e10 for us)
- I have expanded the limitations for the function even more:
   - `for each displacement there can be only one force` means that it is not possible to draw vertical lines - when you want to read the graph you insert displacement and you want to get 1 force back. -> this would not be possible if there is a vertical line since for one displacement there would be more forces to choose from. The minimum distance of two consectuive points in the diagram shouldn't be less than 1e-6. 
   ![image](https://user-images.githubusercontent.com/89633938/220928427-070f74ff-ad99-40b6-982d-a03ba8fe777e.png)

 
   - I also added the condition that both values one point of the function must be either positive or negative. It is not possible positive displacement to cause negative force. (There can be lines drawn into bottom left and top right corner)
   - And the graph must contain point 0; 0 - it must go through origin
  
   >I think they can be all validated (even though some might be obvious)
